FV3/atmos_cubed_sphere/tools/coarse_grained_diagnostics.F90

```diff
@@ -124,11 +128,71 @@ subroutine fv_coarse_diag_model_levels(Atm, Time)
     endif    
   end subroutine fv_coarse_diag_model_levels
   
+  subroutine fv_coarse_diag_pressure_levels(Atm, Time)
+     type(fv_atmos_type), intent(in), target :: Atm(:)
+     type(time_type), intent(in) :: Time
+     
+     real, allocatable, dimension(:,:,:) :: work_3d_coarse, remapped_field, phalf, upsampled_coarse_phalf, masked_area_weights
+     integer :: diagnostic_ids_3d(Atm(tile_count)%coarse_graining%idiag%n_3d_diagnostics)
+     integer :: area_weighted_diagnostic_ids(1)
+     integer :: is, ie, js, je, is_coarse, ie_coarse, js_coarse, je_coarse, npz
+     logical :: used
+ 
+     call get_diagnostic_ids_3d(Atm(tile_count)%coarse_graining%idiag, diagnostic_ids_3d)
+     call get_area_weighted_diagnostic_ids(Atm(tile_count)%coarse_graining%idiag, area_weighted_diagnostic_ids)
+     call get_fine_array_bounds(is, ie, js, je)
+     call get_coarse_array_bounds(is_coarse, ie_coarse, js_coarse, je_coarse)
+     npz = Atm(tile_count)%npz
+ 
+     if (any(diagnostic_ids_3d > 0)) then
+        allocate(work_3d_coarse(is_coarse:ie_coarse,js_coarse:je_coarse,1:npz))
+        allocate(remapped_field(is:ie,js:je,1:npz))
+        allocate(phalf(is:ie,js:je,1:npz+1))      
+        allocate(upsampled_coarse_phalf(is:ie,js:je,1:npz+1))
+
+        call vertical_remapping_requirements( &
+             Atm(tile_count)%delp(is:ie,js:je,1:npz), &
+             Atm(tile_count)%gridstruct%area(is:ie,js:je), &
+             Atm(tile_count)%ptop, &
+             phalf, &
+             upsampled_coarse_phalf)
+     endif
+
+     if (any(area_weighted_diagnostic_ids > 0)) then
+          allocate(masked_area_weights(is:ie,js:je,1:npz))
+          call mask_area_weights( &
+             Atm(tile_count)%gridstruct%area(is:ie,js:je), &
+             phalf, &
+             upsampled_coarse_phalf, &
+             masked_area_weights)
+     endif
+ 
+     if (Atm(tile_count)%coarse_graining%idiag%id_omega_coarse > 0) then
```

Comment:
Why does this variable get handled specially?

Reply:
The way things are set up now, every coarse diagnostic will be treated specially.  Currently just as a demonstration I'm including an omega diagnostic; I plan on submitting a follow-up PR that fills out this module with a larger set of commonly used diagnostics. 

It would be a bit of a refactor, but in the end might be cleaner if I set things up more like they are in the physics.  That is, define a type that contains the metadata for a diagnostic (e.g. the name, units, id, etc.), a pointer to its data, and a weighting method (e.g. area or mass weighted).  I could then iterate over a list of those types when initializing those diagnostics, and also iterate over them when computing them and saving the data out to disk.  

Do you think that would be worth it?  Maybe I could do the refactor as part of that PR that fills out the diagnostic set?  The example types I'm referring to are defined here:

https://github.com/VulcanClimateModeling/fv3gfs-fortran/blob/9a6dcb8ee627145e0ca84f2e6f5bef2a6fb86467/FV3/gfsphysics/GFS_layer/GFS_diagnostics.F90#L21-L42

The physics loops over a list of these types in [registering](https://github.com/VulcanClimateModeling/fv3gfs-fortran/blob/9a6dcb8ee627145e0ca84f2e6f5bef2a6fb86467/FV3/io/FV3GFS_io.F90#L2412-L2514) and [computing](https://github.com/VulcanClimateModeling/fv3gfs-fortran/blob/9a6dcb8ee627145e0ca84f2e6f5bef2a6fb86467/FV3/io/FV3GFS_io.F90#L2526-L2786) the diagnostics later.